### PR TITLE
RF-12966 `mobile-upload` and `environments` command

### PR DESCRIPTION
**Pull request by shanempope**

### CODE REVIEW

File: rainforest/mobile_upload.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	if err != nil {
+		<-errChan
+		return err
+	}
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+	req.ContentLength = contentLength
+
+	// Perform the upload
+	client := &http.Client{}
+	resp, err := client.Do(req)
```
Comment (epaulet) [inline rainforest/mobile_upload.go]: The other goroutine needs to complete writing before we send this request, right? I think we have to way on `<-errChan` before this.

File: rainforest/mobile_upload.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+		<-errChan
+		if err != nil {
+			return err
+		}
+
+		return fmt.Errorf("There was an error uploading your file - %v: %v", fileName, string(body))
+	}
+
+	return <-errChan
```
Comment (epaulet) [inline rainforest/mobile_upload.go]: On a related note, `errChan` should be empty at this point already.

File: rainforest/mobile_upload.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+// SiteEnvironment type represents a single SiteEnvironment returned by the API call for a list of sites
+type SiteEnvironment struct {
+	ID            int    `json:"id"`
+	SiteID        int    `json:"site_id"`
+	EnvironmentID int    `json:"environment_id"`
+	URL           string `json:"url"`
+}
+
+func (c *Client) getSiteEnvironments(siteID int, environmentID int) (int, error) {
```
Comment (epaulet) [inline rainforest/mobile_upload.go]: This should be `getSiteEnvironmentID` or something right? It's only returning the ID. Also do we normally return -1 or 0 alongside errors?

File: mobile_upload.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	err = api.UpdateURL(siteID, environmentID, presignedPostData.RainforestURL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var allowedExtensions = map[string]bool{".apk": true, ".ipa": true, ".zip": true, ".gz": true, ".tar.gz": true}
+var allowedExtensionsPretty = ".apk, .ipa, .zip, .gz, .tar.gz"
```
[the remark this answers is not in the document]
Comment (kruszczynski) [reply mobile_upload.go]: +1

File: rainforest/mobile_upload.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+			return err
+		}
+
+		return fmt.Errorf("There was an error uploading your file - %v: %v", fileName, string(body))
+	}
+
+	return <-errChan
+}
+
+func addFields(requiredFields map[string]string, writer *multipart.Writer) {
```
Comment (kruszczynski) [inline rainforest/mobile_upload.go]: Could we rename this to `writeFields`?

File: rainforest/test/testfile.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+testing123
```
[the remark this answers is not in the document]
Comment (shanempope) [reply rainforest/test/testfile.txt]: Just a test file used in a test

File: rainforest/mobile_upload.go (hunk 1 above)
Comment (shanempope) [reply rainforest/mobile_upload.go]: That `Do` is what uses the pipe to stream the request and not load it into memory.

File: rainforest/mobile_upload.go (hunk 3 above)
Comment (shanempope) [reply rainforest/mobile_upload.go]: I see we return 0 in a few other places so switched it to that.

File: mobile_upload.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	if !allowedExtensions[fileExt] {
+		return cli.NewExitError(fmt.Sprintf("Invalid file extension. - %v. Allowed Extensions: %v", fileExt, allowedExtensionsPretty), 1)
+	}
+
+	siteID, err := strconv.Atoi(c.String("site-id"))
+	if err != nil {
+		return cli.NewExitError("site-id flag required", 1)
+	}
+
+	environmentID, err := strconv.Atoi(c.String("environment-id"))
```
[the remark this answers is not in the document]
Comment (shanempope) [reply mobile_upload.go]: yes

File: rainforest/mobile_upload.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+		// Add the required fields from S3
+		addFields(postData.RequiredFields, writer)
+
+		part, err := writer.CreateFormFile("file", fileName)
+		if err != nil {
+			errChan <- err
+			return
+		}
+		if _, err := io.CopyN(part, f, fileSize); err != nil {
```
[the remark this answers is not in the document]
Comment (shanempope) [reply rainforest/mobile_upload.go]: The writer is a pipe so I need to copy the full length of the file.

File: mobile_upload.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+	err = api.UpdateURL(siteID, environmentID, presignedPostData.RainforestURL)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func allowedExtensions(extension string) bool {
```
Comment (epaulet) [inline mobile_upload.go]: I would name this `isAllowedExtension` or something along those lines.

File: mobile_upload.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		return cli.NewExitError("environment-id flag required", 1)
+	}
+
+	// Open app and return early with an error if we fail
+	f, err := os.Open(filePath)
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+	defer f.Close()
```
[the remark this answers is not in the document]
Comment (epaulet) [reply mobile_upload.go]: This shouldn't be an issue based on https://golang.org/pkg/os/#Open

**Pull request merged**